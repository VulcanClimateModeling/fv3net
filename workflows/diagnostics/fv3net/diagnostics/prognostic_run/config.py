from typing import Mapping, List
import intake


def get_verification_entries(
    name: str, catalog: intake.catalog.Catalog, evaluation_grid="c48"
) -> Mapping[str, List[str]]:
    """Given simulation name, return catalog keys for c48 dycore and physics data.

    Args:
        name: Simulation to use for verification.
        catalog: Catalog to search for verification data.
        evaluation_grid: Grid upon which to compute diagnostics

    Returns:
        Mapping from category name ('physics', 'dycore', or '3d') to sequence
        of catalog keys representing given diagnostics for specified simulation.
    """
    entries: Mapping[str, List[str]] = {"2d": [], "3d": []}
    for item in catalog:
        metadata = catalog[item].metadata
        item_simulation = metadata.get("simulation", None)
        item_grid = metadata.get("grid", None)
        item_category = metadata.get("category", None)

        if item_simulation == name and item_grid == evaluation_grid:
            if item_category is not None:
                entries[item_category].append(item)

    if len(entries["2d"]) == 0:
        raise ValueError(
            f"No {evaluation_grid} 2d diagnostics found in catalog for "
            f"simulation {name}."
        )

    return entries
