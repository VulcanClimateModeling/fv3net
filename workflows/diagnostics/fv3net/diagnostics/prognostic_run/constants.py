from typing import Mapping, Sequence


MovieUrls = Mapping[str, Sequence[str]]

# Added to atmos diags by fortran model if interval-averaged.
# Removed before coarsening run data to avoid errors.
FORTRAN_TILE_ONLY_VARS = ("average_T1", "average_T2", "average_DT")

# Renamed keys already have "_coarse" suffix removed prior to the renaming being applied
VERIFICATION_RENAME_MAP = {
    "40day_may2020": {"2d": {"TB": "TMPlowest", "tsfc": "TMPsfc"}},
    "1yr_pire_postspinup": {"2d": {"TB": "TMPlowest", "tsfc": "TMPsfc"}},
}

RMSE_VARS = [
    "UGRDlowest",
    "VGRDlowest",
    "UGRD850",
    "UGRD200",
    "VGRD850",
    "VGRD200",
    "VORT850",
    "VORT200",
    "TMP500_300",
    "TMPlowest",
    "TMP850",
    "TMP500",
    "TMP200",
    "h500",
    "q500",
    "PRMSL",
    "PRESsfc",
    "PWAT",
    "water_vapor_path",
    "VIL",
    "iw",
]

GLOBAL_AVERAGE_DYCORE_VARS = [
    "UGRD850",
    "UGRD200",
    "TMP500_300",
    "TMPlowest",
    "TMP850",
    "TMP200",
    "w500",
    "h500",
    "RH1000",
    "RH850",
    "RH500",
    "RH200",
    "q1000",
    "q500",
    "q200",
    "PRMSL",
    "PRESsfc",
    "PWAT",
    "water_vapor_path",
    "VIL",
    "iw",
    "VGRD850",
]

GLOBAL_AVERAGE_PHYSICS_VARS = [
    "column_integrated_pQ1",
    "column_integrated_dQ1",
    "column_integrated_nQ1",
    "column_integrated_Q1",
    "column_integrated_pQ2",
    "column_integrated_dQ2",
    "column_integrated_nQ2",
    "column_integrated_Q2",
    "column_int_dQu",
    "column_int_dQv",
    "total_precip_to_surface",
    "PRATEsfc",
    "LHTFLsfc",
    "SHTFLsfc",
    "DSWRFsfc",
    "DLWRFsfc",
    "USWRFtoa",
    "ULWRFtoa",
    "TMPsfc",
    "UGRD10m",
    "MAXWIND10m",
    "SOILM",
    "VGRD10m",
    "is_novelty",
    "centered_score",
]

GLOBAL_AVERAGE_VARS = GLOBAL_AVERAGE_DYCORE_VARS + GLOBAL_AVERAGE_PHYSICS_VARS

GLOBAL_BIAS_PHYSICS_VARS = [
    "column_integrated_Q1",
    "column_integrated_Q2",
    "total_precip_to_surface",
    "LHTFLsfc",
    "SHTFLsfc",
    "USWRFtoa",
    "ULWRFtoa",
    "TMPsfc",
    "UGRD10m",
    "MAXWIND10m",
    "SOILM",
    "VGRD10m",
]

GLOBAL_BIAS_VARS = GLOBAL_AVERAGE_DYCORE_VARS + GLOBAL_BIAS_PHYSICS_VARS

DIURNAL_CYCLE_VARS = [
    "column_integrated_dQ1",
    "column_integrated_nQ1",
    "column_integrated_pQ1",
    "column_integrated_Q1",
    "column_integrated_dQ2",
    "column_integrated_nQ2",
    "column_integrated_pQ2",
    "column_integrated_Q2",
    "total_precip_to_surface",
    "PRATEsfc",
    "LHTFLsfc",
]

TIME_MEAN_VARS = [
    "column_integrated_Q1",
    "column_integrated_Q2",
    "total_precip_to_surface",
    "UGRD850",
    "UGRD200",
    "TMPsfc",
    "TMP850",
    "TMP200",
    "h500",
    "w500",
    "PRESsfc",
    "PWAT",
    "water_vapor_path",
    "LHTFLsfc",
    "SHTFLsfc",
    "DSWRFsfc",
    "DLWRFsfc",
    "USWRFtoa",
    "ULWRFtoa",
]

PRESSURE_INTERPOLATED_VARS = [
    "air_temperature",
    "specific_humidity",
    "relative_humidity",
    "eastward_wind",
    "northward_wind",
    "vertical_wind",
    "dQ1",
    "dQ2",
    "dQu",
    "dQv",
    "tendency_of_air_temperature_due_to_machine_learning",
    "tendency_of_specific_humidity_due_to_machine_learning",
    "air_temperature_tendency_due_to_nudging",
    "specific_humidity_tendency_due_to_nudging",
]

PRECIP_RATE = "total_precip_to_surface"
MASS_STREAMFUNCTION_MID_TROPOSPHERE = (
    "mass_streamfunction_300_700_time_dependent_zonal_mean"
)
MASS_STREAMFUNCTION_MID_TROPOSPHERE_TIME_MEAN = (
    "mass_streamfunction_300_700_zonal_and_time_mean"
)

PERCENTILES = [25, 50, 75, 90, 99, 99.9]
TOP_LEVEL_METRICS = {
    "rmse_5day": ["h500", "tmp850"],
    "rmse_of_time_mean": [PRECIP_RATE, "pwat", "tmp200"],
    "time_and_land_mean_bias": [PRECIP_RATE],
    "rmse_of_time_mean_land": ["tmp850"],
    "tropics_max_minus_min": [MASS_STREAMFUNCTION_MID_TROPOSPHERE_TIME_MEAN],
    "tropical_ascent_region_mean": ["column_integrated_q1", "water_vapor_path"],
}

GRID_VARS = ("lat", "latb", "lon", "lonb")
GRID_INTERFACE_COORDS = ("x_interface", "y_interface")

# https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.figure.html
DEFAULT_FIGURE_WIDTH = 6.4
DEFAULT_FIGURE_HEIGHT = 4.8

# For inferring a readable aspect ratio for tropical Hovmoller plots
REFERENCE_HOVMOLLER_DURATION_SECONDS = 40 * 86400
