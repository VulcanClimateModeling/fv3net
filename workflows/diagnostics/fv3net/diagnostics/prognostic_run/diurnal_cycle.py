import logging
import numpy as np
import xarray as xr
import flox.xarray

import vcm

logger = logging.getLogger(__name__)

SECONDS_PER_DAY = 86400


def calc_diagnostics(prognostic, verification, grid):
    """
    Diagnostic function for calculating all diurnal cycle
    information from physics output.
    """

    logger.info("Preparing diurnal cycle diagnostics")

    prog_diurnal_ds = _calc_ds_diurnal_cycle(prognostic.assign(lon=grid["lon"]))
    verif_diurnal_ds = _calc_ds_diurnal_cycle(verification.assign(lon=grid["lon"]))

    prog_diurnal_ds = _add_diurnal_moisture_components(prog_diurnal_ds)
    verif_diurnal_ds = _add_diurnal_moisture_components(verif_diurnal_ds)

    prog_diurnal_ds = _add_diurn_bias(prog_diurnal_ds, verif_diurnal_ds)

    return prog_diurnal_ds


def _calc_ds_diurnal_cycle(ds):
    """
    Calculates the diurnal cycle for all variables.  Expects
    time dimension and longitude variable "lon".

    Uses flox.xarray.xarray_reduce to speed up the diurnal cycle calc
    See https://xarray.dev/blog/flox
    """
    local_time = vcm.local_time(ds, time="time", lon_var="lon")
    local_time.attrs = {"long_name": "local time", "units": "hour"}

    local_time = np.floor(local_time).load() # equivalent to hourly binning
    labels = np.unique(local_time)
    
    diurnal_cycles = xr.Dataset()
    for var in ds.data_vars:
        logger.info(f"Calculating diurnal cycle for {var}")
        with xr.set_options(keep_attrs=True):
            diurnal_cycles[var] = flox.xarray.xarray_reduce(
                ds[var], 
                local_time.rename("local_time"),
                func="nanmean", 
                expected_groups=(labels,),
                method="map-reduce",
                engine="flox"
            ).compute()
    return diurnal_cycles


def _add_diurnal_moisture_components(diurnal_cycles: xr.Dataset):
    """
    Add individual moisture components for diurnal cycle plots with a long-name
    and attributes.  The naming is used by report generation to determine the
    component shorthand.  E.g., diurn_component_<component_name>
    """
    if "surf_evap" in diurnal_cycles.variables:
        evap = diurnal_cycles.surf_evap
    else:
        evap = vcm.latent_heat_flux_to_evaporation(diurnal_cycles["LHTFLsfc"])
    evap *= SECONDS_PER_DAY
    evap.attrs = {"long_name": "Evaporation", "units": "mm/day"}
    diurnal_cycles["diurn_component_evaporation"] = evap

    precip = diurnal_cycles["PRATEsfc"] * SECONDS_PER_DAY
    precip.attrs = {"long_name": "Physics precipitation", "units": "mm/day"}
    diurnal_cycles["diurn_component_physics-precipitation"] = precip

    dQ2 = diurnal_cycles["column_integrated_dQ2"]
    diurnal_cycles["diurn_component_<-dQ2>"] = -dQ2
    diurnal_cycles["diurn_component_<-dQ2>"].attrs = {
        "long_name": "<-dQ2> column integrated drying from ML",
        "units": "mm/day",
    }

    precip_phys_ml_nudging = diurnal_cycles["total_precip_to_surface"]
    precip_phys_ml_nudging.attrs = {
        "long_name": "Total precip to surface (max(PRATE-<dQ2>+<nQ2>, 0))",
        "units": "mm/day",
    }
    diurnal_cycles["diurn_component_total-precipitation"] = precip_phys_ml_nudging

    return diurnal_cycles


def _add_diurn_bias(prognostic_diurnal, verif_diurnal):
    """
    Add comparisons of diurnal cycle against verification data for plotting
    """

    evap_compare = (
        prognostic_diurnal["diurn_component_evaporation"]
        - verif_diurnal["diurn_component_evaporation"]
    )
    evap_compare.attrs = {
        "long_name": "Evaporation diurnal cycle bias [run - verif]",
        "units": "mm/day",
    }
    prognostic_diurnal["diurn_bias_evaporation"] = evap_compare

    precip_compare = (
        prognostic_diurnal["diurn_component_total-precipitation"]
        - verif_diurnal["diurn_component_total-precipitation"]
    )
    precip_compare.attrs = {
        "long_name": (
            "Total precip to surface (max(PRATE-<dQ2>-<nQ2>, 0)) "
            "diurnal cycle bias [run - verif]"
        ),
        "units": "mm/day",
    }
    prognostic_diurnal["diurn_bias_total-precipitation"] = precip_compare

    net_precip_compare = precip_compare - evap_compare
    net_precip_compare.attrs = {
        "long_name": ("Net precip (-<Q2>) diurnal cycle bias [run - verif]"),
        "units": "mm/day",
    }
    prognostic_diurnal["net_precip_against_verif"] = net_precip_compare

    return prognostic_diurnal
