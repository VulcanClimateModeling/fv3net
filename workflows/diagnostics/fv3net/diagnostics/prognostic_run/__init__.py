from .config import get_verification_entries
from .load_run_data import load_dycore, load_physics


__all__ = dir()
