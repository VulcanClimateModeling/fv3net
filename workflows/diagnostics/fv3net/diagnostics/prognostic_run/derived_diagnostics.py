from typing import Sequence, Tuple

import xarray as xr

import vcm

from fv3net.diagnostics._shared.registry import Registry
from .constants import WVP, COL_DRYING


def merge_derived(diags: Sequence[Tuple[str, xr.DataArray]]) -> xr.Dataset:
    out = xr.Dataset()
    for name, da in diags:
        if len(da.dims) != 0:
            # don't add empty DataArrays, which are returned in case of missing inputs
            out[name] = da
    return out


# all functions added to this registry must take a single xarray Dataset as
# input and return a single xarray DataArray
derived_registry = Registry(merge_derived)


@derived_registry.register("mass_streamfunction_pressure_level_zonal_time_mean")
def psi_value(diags: xr.Dataset) -> xr.DataArray:
    if "northward_wind_pressure_level_zonal_time_mean" not in diags:
        return xr.DataArray()
    northward_wind = diags["northward_wind_pressure_level_zonal_time_mean"]
    return vcm.mass_streamfunction(northward_wind)


@derived_registry.register("mass_streamfunction_pressure_level_zonal_bias")
def psi_bias(diags: xr.Dataset) -> xr.DataArray:
    if "northward_wind_pressure_level_zonal_bias" not in diags:
        return xr.DataArray()
    northward_wind_bias = diags["northward_wind_pressure_level_zonal_bias"]
    return vcm.mass_streamfunction(northward_wind_bias)


@derived_registry.register("mass_streamfunction_300_700_zonal_and_time_mean")
def psi_value_mid_troposphere(diags: xr.Dataset) -> xr.DataArray:
    if "northward_wind_pressure_level_zonal_time_mean" not in diags:
        return xr.DataArray()
    northward_wind = diags["northward_wind_pressure_level_zonal_time_mean"]
    psi = vcm.mass_streamfunction(northward_wind).sel(pressure=slice(30000, 70000))
    psi_mid_trop = psi.weighted(psi.pressure).mean("pressure")
    return psi_mid_trop.assign_attrs(
        long_name="mass streamfunction 300-700hPa average", units="Gkg/s"
    )


@derived_registry.register("mass_streamfunction_300_700_zonal_bias")
def psi_bias_mid_troposphere(diags: xr.Dataset) -> xr.DataArray:
    if "northward_wind_pressure_level_zonal_bias" not in diags:
        return xr.DataArray()
    northward_wind_bias = diags["northward_wind_pressure_level_zonal_bias"]
    psi = vcm.mass_streamfunction(northward_wind_bias).sel(pressure=slice(30000, 70000))
    psi_mid_trop = psi.weighted(psi.pressure).mean("pressure")
    return psi_mid_trop.assign_attrs(
        long_name="mass streamfunction 300-700hPa average", units="Gkg/s"
    )


@derived_registry.register(f"conditional_average_of_{COL_DRYING}_on_{WVP}")
def conditional_average(diags: xr.Dataset) -> xr.DataArray:
    count_name = f"{WVP}_versus_{COL_DRYING}_hist_2d"
    q2_bin_name = f"{COL_DRYING}_bins"
    q2_bin_widths_name = f"{COL_DRYING}_bin_width_hist_2d"
    if count_name not in diags:
        return xr.DataArray()
    count = diags[count_name]
    q2_bin_centers = diags[q2_bin_name] + 0.5 * diags[q2_bin_widths_name]
    average = vcm.weighted_average(q2_bin_centers, count, dims=[q2_bin_name])
    return average.assign_attrs(
        long_name="Conditional average of -<Q2> on water vapor path", units="mm/day"
    )
