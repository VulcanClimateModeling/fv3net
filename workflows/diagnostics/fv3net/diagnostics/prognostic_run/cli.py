import argparse
from fv3net.diagnostics.prognostic_run import metrics, compute
from fv3net.diagnostics.prognostic_run.views import movies, static_report
from fv3net.diagnostics.prognostic_run.apps import log_viewer


def dissoc(namespace, key):
    return argparse.Namespace(
        **{k: val for k, val in vars(namespace).items() if k != key}
    )


def get_parser():
    parser = argparse.ArgumentParser(description="Prognostic run diagnostics")
    subparsers = parser.add_subparsers(
        help="Prognostic run diagnostics", required=True, dest="command"
    )

    compute.register_parser(subparsers)
    metrics.register_parser(subparsers)
    movies.register_parser(subparsers)
    static_report.register_parser(subparsers)
    log_viewer.register_parser(subparsers)
    return parser


def main():
    parser = get_parser()
    args = parser.parse_args()

    # need to remove the 'func' entry since some of these scripts save these
    # arguments in netCDF attributes
    args.func(dissoc(args, "func"))


if __name__ == "__main__":
    main()
