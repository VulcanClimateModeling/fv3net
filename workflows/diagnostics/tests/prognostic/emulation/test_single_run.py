import numpy as np
import pytest
import vcm

from fv3net.diagnostics.prognostic_run.emulation import single_run

cdl = """
netcdf out {
dimensions:
    time = 3;
    tile = 6;
    z = 79 ;
    y = 12 ;
    x = 12 ;
    phalf = 80 ;
    y_interface = 13 ;
    x_interface = 13 ;
variables:
    double time(time) ;
        time:_FillValue = NaN ;
        time:calendar_type = "JULIAN" ;
        time:cartesian_axis = "T" ;
        time:long_name = "time" ;
        time:units = "hours since 2016-06-11" ;
        time:calendar = "JULIAN" ;
    double z(z) ;
        z:_FillValue = NaN ;
        z:cartesian_axis = "Z" ;
        z:edges = "phalf" ;
        z:long_name = "ref full pressure level" ;
        z:positive = "down" ;
        z:units = "mb" ;
    double y(y) ;
        y:_FillValue = NaN ;
        y:cartesian_axis = "Y" ;
        y:long_name = "T-cell latitude" ;
        y:units = "degrees_N" ;
    double x(x) ;
        x:_FillValue = NaN ;
        x:cartesian_axis = "X" ;
        x:long_name = "T-cell longitude" ;
        x:units = "degrees_E" ;
    float delp(time, tile, z, y, x) ;
        delp:_FillValue = NaNf ;
        delp:cell_methods = "time: point" ;
        delp:long_name = "pressure thickness" ;
        delp:units = "pa" ;
        delp:coordinates = "time" ;
    double phalf(phalf) ;
        phalf:_FillValue = NaN ;
        phalf:cartesian_axis = "Z" ;
        phalf:long_name = "ref half pressure level" ;
        phalf:positive = "down" ;
        phalf:units = "mb" ;
    float surface_precipitation_due_to_zhao_carr_emulator(time, tile, y, x) ;
        surface_precipitation_due_to_zhao_carr_emulator:_FillValue = NaNf ;
        surface_precipitation_due_to_zhao_carr_emulator:cell_methods = "time: point" ;
        surface_precipitation_due_to_zhao_carr_emulator:long_name = "surface precipitation due to zhao_carr_microphysics emulator" ;
        surface_precipitation_due_to_zhao_carr_emulator:units = "kg/m^2/s" ;
        surface_precipitation_due_to_zhao_carr_emulator:coordinates = "time" ;
    float surface_precipitation_due_to_zhao_carr_physics(time, tile, y, x) ;
        surface_precipitation_due_to_zhao_carr_physics:_FillValue = NaNf ;
        surface_precipitation_due_to_zhao_carr_physics:cell_methods = "time: point" ;
        surface_precipitation_due_to_zhao_carr_physics:long_name = "surface precipitation due to zhao_carr_microphysics physics" ;
        surface_precipitation_due_to_zhao_carr_physics:units = "kg/m^2/s" ;
        surface_precipitation_due_to_zhao_carr_physics:coordinates = "time" ;
    float tendency_of_air_temperature_due_to_gscond_emulator(time, tile, z, y, x) ;
        tendency_of_air_temperature_due_to_gscond_emulator:_FillValue = NaNf ;
        tendency_of_air_temperature_due_to_gscond_emulator:cell_methods = "time: point" ;
        tendency_of_air_temperature_due_to_gscond_emulator:long_name = "temperature tendency due to zhao_carr_gscond emulator" ;
        tendency_of_air_temperature_due_to_gscond_emulator:units = "K/s" ;
        tendency_of_air_temperature_due_to_gscond_emulator:coordinates = "time" ;
    float tendency_of_air_temperature_due_to_gscond_physics(time, tile, z, y, x) ;
        tendency_of_air_temperature_due_to_gscond_physics:_FillValue = NaNf ;
        tendency_of_air_temperature_due_to_gscond_physics:cell_methods = "time: point" ;
        tendency_of_air_temperature_due_to_gscond_physics:long_name = "temperature tendency due to zhao_carr_gscond physics" ;
        tendency_of_air_temperature_due_to_gscond_physics:units = "K/s" ;
        tendency_of_air_temperature_due_to_gscond_physics:coordinates = "time" ;
    float tendency_of_air_temperature_due_to_zhao_carr_emulator(time, tile, z, y, x) ;
        tendency_of_air_temperature_due_to_zhao_carr_emulator:_FillValue = NaNf ;
        tendency_of_air_temperature_due_to_zhao_carr_emulator:cell_methods = "time: point" ;
        tendency_of_air_temperature_due_to_zhao_carr_emulator:long_name = "temperature tendency due to zhao_carr_microphysics emulator" ;
        tendency_of_air_temperature_due_to_zhao_carr_emulator:units = "K/s" ;
        tendency_of_air_temperature_due_to_zhao_carr_emulator:coordinates = "time" ;
    float tendency_of_air_temperature_due_to_zhao_carr_physics(time, tile, z, y, x) ;
        tendency_of_air_temperature_due_to_zhao_carr_physics:_FillValue = NaNf ;
        tendency_of_air_temperature_due_to_zhao_carr_physics:cell_methods = "time: point" ;
        tendency_of_air_temperature_due_to_zhao_carr_physics:long_name = "temperature tendency due to zhao_carr_microphysics physics" ;
        tendency_of_air_temperature_due_to_zhao_carr_physics:units = "K/s" ;
        tendency_of_air_temperature_due_to_zhao_carr_physics:coordinates = "time" ;
    float tendency_of_cloud_water_due_to_gscond_physics(time, tile, z, y, x) ;
        tendency_of_cloud_water_due_to_gscond_physics:_FillValue = NaNf ;
        tendency_of_cloud_water_due_to_gscond_physics:cell_methods = "time: point" ;
        tendency_of_cloud_water_due_to_gscond_physics:long_name = "cloud water due to zhao_carr_gscond physics" ;
        tendency_of_cloud_water_due_to_gscond_physics:units = "kg/kg/s" ;
        tendency_of_cloud_water_due_to_gscond_physics:coordinates = "time" ;
    float tendency_of_cloud_water_due_to_zhao_carr_emulator(time, tile, z, y, x) ;
        tendency_of_cloud_water_due_to_zhao_carr_emulator:_FillValue = NaNf ;
        tendency_of_cloud_water_due_to_zhao_carr_emulator:cell_methods = "time: point" ;
        tendency_of_cloud_water_due_to_zhao_carr_emulator:long_name = "cloud water due to zhao_carr_microphysics emulator" ;
        tendency_of_cloud_water_due_to_zhao_carr_emulator:units = "kg/kg/s" ;
        tendency_of_cloud_water_due_to_zhao_carr_emulator:coordinates = "time" ;
    float tendency_of_cloud_water_due_to_zhao_carr_physics(time, tile, z, y, x) ;
        tendency_of_cloud_water_due_to_zhao_carr_physics:_FillValue = NaNf ;
        tendency_of_cloud_water_due_to_zhao_carr_physics:cell_methods = "time: point" ;
        tendency_of_cloud_water_due_to_zhao_carr_physics:long_name = "cloud water due to zhao_carr_microphysics physics" ;
        tendency_of_cloud_water_due_to_zhao_carr_physics:units = "kg/kg/s" ;
        tendency_of_cloud_water_due_to_zhao_carr_physics:coordinates = "time" ;
    float tendency_of_specific_humidity_due_to_gscond_emulator(time, tile, z, y, x) ;
        tendency_of_specific_humidity_due_to_gscond_emulator:_FillValue = NaNf ;
        tendency_of_specific_humidity_due_to_gscond_emulator:cell_methods = "time: point" ;
        tendency_of_specific_humidity_due_to_gscond_emulator:long_name = "specific humidity tendency due to zhao_carr_gscond emulator" ;
        tendency_of_specific_humidity_due_to_gscond_emulator:units = "kg/kg/s" ;
        tendency_of_specific_humidity_due_to_gscond_emulator:coordinates = "time" ;
    float tendency_of_specific_humidity_due_to_gscond_physics(time, tile, z, y, x) ;
        tendency_of_specific_humidity_due_to_gscond_physics:_FillValue = NaNf ;
        tendency_of_specific_humidity_due_to_gscond_physics:cell_methods = "time: point" ;
        tendency_of_specific_humidity_due_to_gscond_physics:long_name = "specific humidity tendency due to zhao_carr_gscond physics" ;
        tendency_of_specific_humidity_due_to_gscond_physics:units = "kg/kg/s" ;
        tendency_of_specific_humidity_due_to_gscond_physics:coordinates = "time" ;
    float tendency_of_specific_humidity_due_to_zhao_carr_emulator(time, tile, z, y, x) ;
        tendency_of_specific_humidity_due_to_zhao_carr_emulator:_FillValue = NaNf ;
        tendency_of_specific_humidity_due_to_zhao_carr_emulator:cell_methods = "time: point" ;
        tendency_of_specific_humidity_due_to_zhao_carr_emulator:long_name = "specific humidity tendency due to zhao_carr_microphysics emulator" ;
        tendency_of_specific_humidity_due_to_zhao_carr_emulator:units = "kg/kg/s" ;
        tendency_of_specific_humidity_due_to_zhao_carr_emulator:coordinates = "time" ;
    float tendency_of_specific_humidity_due_to_zhao_carr_physics(time, tile, z, y, x) ;
        tendency_of_specific_humidity_due_to_zhao_carr_physics:_FillValue = NaNf ;
        tendency_of_specific_humidity_due_to_zhao_carr_physics:cell_methods = "time: point" ;
        tendency_of_specific_humidity_due_to_zhao_carr_physics:long_name = "specific humidity tendency due to zhao_carr_microphysics physics" ;
        tendency_of_specific_humidity_due_to_zhao_carr_physics:units = "kg/kg/s" ;
        tendency_of_specific_humidity_due_to_zhao_carr_physics:coordinates = "time" ;
    float area(tile, y, x) ;
        area:_FillValue = NaNf ;
        area:cell_methods = "time: point" ;
        area:long_name = "cell area" ;
        area:units = "m**2" ;
        area:coordinates = "time" ;
    float lat(tile, y, x) ;
        lat:_FillValue = NaNf ;
        lat:cell_methods = "time: point" ;
        lat:long_name = "latitude" ;
        lat:units = "degrees_N" ;
        lat:coordinates = "time" ;
    float latb(tile, y_interface, x_interface) ;
        latb:_FillValue = NaNf ;
        latb:cell_methods = "time: point" ;
        latb:long_name = "latitude" ;
        latb:units = "degrees_N" ;
        latb:coordinates = "time" ;
    float lon(tile, y, x) ;
        lon:_FillValue = NaNf ;
        lon:cell_methods = "time: point" ;
        lon:long_name = "longitude" ;
        lon:units = "degrees_E" ;
        lon:coordinates = "time" ;
    float lonb(tile, y_interface, x_interface) ;
        lonb:_FillValue = NaNf ;
        lonb:cell_methods = "time: point" ;
        lonb:long_name = "longitude" ;
        lonb:units = "degrees_E" ;
        lonb:coordinates = "time" ;
    double x_interface(x_interface) ;
        x_interface:_FillValue = NaN ;
        x_interface:cartesian_axis = "X" ;
        x_interface:long_name = "cell corner longitude" ;
        x_interface:units = "degrees_E" ;
    double y_interface(y_interface) ;
        y_interface:_FillValue = NaN ;
        y_interface:cartesian_axis = "Y" ;
        y_interface:long_name = "cell corner latitude" ;
        y_interface:units = "degrees_N" ;
    double air_temperature(time, tile, z, y, x) ;
        air_temperature:_FillValue = NaN ;
        air_temperature:units = "degK" ;
        air_temperature:coordinates = "time" ;
    double cloud_water_mixing_ratio(time, tile, z, y, x) ;
        cloud_water_mixing_ratio:_FillValue = NaN ;
        cloud_water_mixing_ratio:units = "kg/kg" ;
        cloud_water_mixing_ratio:coordinates = "time" ;
    double eastward_wind(time, tile, z, y, x) ;
        eastward_wind:_FillValue = NaN ;
        eastward_wind:units = "m/s" ;
        eastward_wind:coordinates = "time" ;
    double land_sea_mask(time, tile, y, x) ;
        land_sea_mask:_FillValue = NaN ;
        land_sea_mask:units = "" ;
        land_sea_mask:coordinates = "time" ;
    double latitude(time, tile, y, x) ;
        latitude:_FillValue = NaN ;
        latitude:units = "radians" ;
        latitude:coordinates = "time" ;
    double longitude(time, tile, y, x) ;
        longitude:_FillValue = NaN ;
        longitude:units = "radians" ;
        longitude:coordinates = "time" ;
    double northward_wind(time, tile, z, y, x) ;
        northward_wind:_FillValue = NaN ;
        northward_wind:units = "m/s" ;
        northward_wind:coordinates = "time" ;
    double pressure_thickness_of_atmospheric_layer(time, tile, z, y, x) ;
        pressure_thickness_of_atmospheric_layer:_FillValue = NaN ;
        pressure_thickness_of_atmospheric_layer:units = "Pa" ;
        pressure_thickness_of_atmospheric_layer:coordinates = "time" ;
    double specific_humidity(time, tile, z, y, x) ;
        specific_humidity:_FillValue = NaN ;
        specific_humidity:units = "kg/kg" ;
        specific_humidity:coordinates = "time" ;
    double surface_pressure(time, tile, y, x) ;
        surface_pressure:_FillValue = NaN ;
        surface_pressure:units = "Pa" ;
        surface_pressure:coordinates = "time" ;
    double total_precipitation(time, tile, y, x) ;
        total_precipitation:_FillValue = NaN ;
        total_precipitation:units = "m" ;
        total_precipitation:coordinates = "time" ;
    double vertical_wind(time, tile, z, y, x) ;
        vertical_wind:_FillValue = NaN ;
        vertical_wind:units = "m/s" ;
        vertical_wind:coordinates = "time" ;
data:
    time = 3, 6, 9, 12, 15;
}

"""  # noqa


@pytest.mark.parametrize(
    "func",
    [
        single_run.plot_histogram_begin_end,
        # this test fails in CI for some reason
        # single_run.plot_cloud_weighted_average,
        single_run.plot_cloud_maps,
        single_run.skill_table,
        single_run.skill_time_table,
        single_run.log_lat_vs_p_skill("cloud_water"),
    ],
)
def test_log_functions(func):

    ds = vcm.cdl_to_dataset(cdl)
    for key in ds:
        ds[key].values[:] = 0

    for key in set(ds.coords) - {"time"}:
        ds[key].values[:] = np.arange(len(ds[key]))

    nx = len(ds.x)
    ds["lat"].values[:] = np.linspace(-45, 45, nx)
    func(ds)


def test_skill_table(regtest):
    ds = vcm.cdl_to_dataset(cdl)
    output = single_run.skill_table(ds)
    for name in sorted(output):
        print(name, ":", output[name].columns, file=regtest)


def test_compute_summaries(regtest):
    ds = vcm.cdl_to_dataset(cdl)
    output = single_run.compute_summaries(ds)
    print(sorted(output), file=regtest)
