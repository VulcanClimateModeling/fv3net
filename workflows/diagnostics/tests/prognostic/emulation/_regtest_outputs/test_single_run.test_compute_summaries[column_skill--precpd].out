['column_skill/precpd/air_temperature', 'column_skill/precpd/cloud_water', 'column_skill/precpd/specific_humidity']
